tests/bluetooth/bsim_bt/bsim_test_l2cap/src/main_l2cap_ecred.c

```diff
@@ -0,0 +1,563 @@
+#define UNSET_FLAG(flag) (void)atomic_set(&flag, (atomic_t)false)
+#define WAIT_FOR_FLAG_SET(flag)		   \
+	while (!(bool)atomic_get(&flag)) { \
+		(void)k_sleep(K_MSEC(1));  \
+	}
+#define WAIT_FOR_FLAG_UNSET(flag)	  \
+	while ((bool)atomic_get(&flag)) { \
+		(void)k_sleep(K_MSEC(1)); \
+	}
+
+
+#define WAIT_TIME (2400000000) /* 240 seconds*/
+
+#define FAIL(...)				       \
+	do {					       \
+		bst_result = Failed;		       \
+		bs_trace_error_time_line(__VA_ARGS__); \
+	} while (0)
+
+#define PASS(...)				    \
+	do {					    \
+		bst_result = Passed;		    \
+		bs_trace_info_time(1, __VA_ARGS__); \
+	} while (0)
+
+static struct bt_conn *default_conn;
+
+static const struct bt_data ad[] = {
+	BT_DATA_BYTES(BT_DATA_FLAGS, (BT_LE_AD_GENERAL | BT_LE_AD_NO_BREDR)),
+};
+
+#define DATA_MTU 2000
+#define DATA_MPS 65
+#define DATA_BUF_SIZE BT_L2CAP_SDU_BUF_SIZE(DATA_MTU)
+#define L2CAP_CHANNELS 2
+#define SERVERS 1
+#define SDU_SEND_COUNT 200
+
+NET_BUF_POOL_FIXED_DEFINE(rx_data_pool, L2CAP_CHANNELS, BT_L2CAP_BUF_SIZE(DATA_BUF_SIZE), 8, NULL);
+NET_BUF_POOL_FIXED_DEFINE(tx_data_pool, L2CAP_CHANNELS + 1, BT_L2CAP_BUF_SIZE(DATA_MTU), 8, NULL);
+
+static struct bt_l2cap_server servers[SERVERS];
+void send_sdu_chan_worker(struct k_work *item);
+
+static struct channel {
+	uint8_t chan_id; /* Internal number that identifies L2CAP channel. */
+	struct bt_l2cap_le_chan le;
+	bool in_use;
+	int sdus_received;
+	int bytes_to_send;
+	int itteration;
+	struct net_buf *buf;
+	struct k_work work;
+	uint8_t payload[DATA_MTU];
+} channels[L2CAP_CHANNELS];
+CREATE_FLAG(is_connected);
+#define MY_STACK_SIZE 512
+#define MY_PRIORITY 5
+
+K_THREAD_STACK_DEFINE(my_stack_area0, MY_STACK_SIZE);
+struct k_work_q my_work_q0;
+K_THREAD_STACK_DEFINE(my_stack_area1, MY_STACK_SIZE);
+struct k_work_q my_work_q1;
+
+static void init_workqs(void)
+{
+	k_work_queue_init(&my_work_q0);
+	k_work_queue_start(&my_work_q0, my_stack_area0,
+			   K_THREAD_STACK_SIZEOF(my_stack_area0), MY_PRIORITY,
+			   NULL);
+
+	k_work_queue_init(&my_work_q1);
+	k_work_queue_start(&my_work_q1, my_stack_area1,
+			   K_THREAD_STACK_SIZEOF(my_stack_area1), MY_PRIORITY,
+			   NULL);
+}
+
+static struct channel *get_free_channel(void)
+{
+	struct channel *chan;
+
+	for (int idx = 0; idx < L2CAP_CHANNELS; idx++) {
+		if (channels[idx].in_use) {
+			continue;
+		}
+		chan = &channels[idx];
+		(void)memset(chan, 0, sizeof(*chan));
+		chan->chan_id = idx;
+		channels[idx].in_use = true;
+		(void)memset(chan->payload, idx, sizeof(chan->payload));
+		k_work_init(&chan->work, send_sdu_chan_worker);
+		return chan;
+	}
+
+	return NULL;
+}
+
+static struct net_buf *chan_alloc_buf_cb(struct bt_l2cap_chan *chan)
+{
+	LOG_DBG("Allocated on chan %p", chan);
+	return net_buf_alloc(&rx_data_pool, K_FOREVER);
+}
+
+static int chan_recv_cb(struct bt_l2cap_chan *l2cap_chan, struct net_buf *buf)
+{
+	struct channel *chan = CONTAINER_OF(l2cap_chan, struct channel, le);
+	int received_iterration = *((int *)(buf->data));
```

Comment:
Suggest to insert the iteration and retrieve the iteration using the `net_buf_pull_le32` and `net_buf_push_le32` functions